src/Parsers/ASTExplainQuery.h

```diff
@@ -1,52 +1,85 @@
 #pragma once
 
-#include <Parsers/IAST.h>
+#include <Parsers/ASTQueryWithOutput.h>
 
 
 namespace DB
 {
 
 
 /// AST, EXPLAIN or other query with meaning of explanation query instead of execution
-class ASTExplainQuery : public IAST
+class ASTExplainQuery : public ASTQueryWithOutput
 {
 public:
     enum ExplainKind
     {
-        ParsedAST,
-        AnalyzedSyntax,
+        ParsedAST, /// 'EXPLAIN AST SELECT ...'
+        AnalyzedSyntax, /// 'EXPLAIN SYNTAX SELECT ...'
+        QueryPlan, /// 'EXPLAIN SELECT ...'
+        QueryPipeline, /// 'EXPLAIN PIPELINE ...'
     };
 
-    ASTExplainQuery(ExplainKind kind_)
-        : kind(kind_)
-    {}
+    ASTExplainQuery(ExplainKind kind_, bool old_syntax_)
+        : kind(kind_), old_syntax(old_syntax_)
+    {
+    }
 
-    String getID(char delim) const override { return "Explain" + (delim + toString(kind)); }
+    String getID(char delim) const override { return "Explain" + (delim + toString(kind, old_syntax)); }
     ExplainKind getKind() const { return kind; }
     ASTPtr clone() const override
     {
         auto res = std::make_shared<ASTExplainQuery>(*this);
         res->children.clear();
         res->children.push_back(children[0]->clone());
+        cloneOutputOptions(*res);
         return res;
     }
 
+    void setExplainedQuery(ASTPtr query_)
+    {
+        children.emplace_back(query_);
+        query = std::move(query_);
+    }
+
+    void setSettings(ASTPtr settings_)
+    {
+        children.emplace_back(settings_);
+        ast_settings = std::move(settings_);
+    }
+
+    const ASTPtr & getExplainedQuery() const { return query; }
+    const ASTPtr & getSettings() const { return ast_settings; }
+
 protected:
-    void formatImpl(const FormatSettings & settings, FormatState & state, FormatStateStacked frame) const override
+    void formatQueryImpl(const FormatSettings & settings, FormatState & state, FormatStateStacked frame) const override
     {
-        settings.ostr << (settings.hilite ? hilite_keyword : "") << toString(kind) << (settings.hilite ? hilite_none : "") << " ";
-        children.at(0)->formatImpl(settings, state, frame);
+        settings.ostr << (settings.hilite ? hilite_keyword : "") << toString(kind, old_syntax) << (settings.hilite ? hilite_none : "");
+
+        if (ast_settings)
+        {
+            settings.ostr << ' ';
+            ast_settings->formatImpl(settings, state, frame);
+        }
+
+        settings.ostr << settings.nl_or_ws;
+        query->formatImpl(settings, state, frame);
     }
 
 private:
     ExplainKind kind;
+    bool old_syntax; /// "EXPLAIN AST" -> "AST", "EXPLAIN SYNTAX" -> "ANALYZE"
+
+    ASTPtr query;
+    ASTPtr ast_settings;
 
-    static String toString(ExplainKind kind)
+    static String toString(ExplainKind kind, bool old_syntax)
```

Comment:
We can safely remove the old `AST` query. It's not for users and no backward compatibility is needed.

Reply:
I agree about  'we don' t care about backward compatibility fot that'. At the same time `Ast` and `analyze` are useful in some (rare) cases (for example to look on rewritten query with joins and lot a of aliases) it if possible it better to keep them.

Reply:
AFAICS It is used only for printing formatted query

- ANALYZE = EXPLAIN SYNTAX
- AST = EXPLAIN AST

Reply:
> ANALYZE = EXPLAIN SYNTAX
AST = EXPLAIN AST

It's correct.

I mainly kept old syntax because it was used in tests. I probably may rewrite all that tests and remove compatibility. I would rather do it in a separate pr.